### PR TITLE
fix: align transitive compatibility checks

### PR DESCRIPTION
**Pull request by davide-armand**
# About this change - What it does

Make so that the transitive compatibility checks that are done in the
schema creation endpoint are also done in the schema validation
endpoint.

In the creation endpoint (`/subjects/<subject-key>/versions`), if the
compatibility mode is transient then the new schema is checked against
all schemas.

In the validation endpoint (`/compatibility/subjects/<subject-key>/versions/<schema-version>`):
 - Before this fix, only the latest schema is checked against (even in
   case of transitive mode)
 - After this fix, in case of transitive mode then all schema are
   checked against.
   Note that in this case the version provided in the POST request
   (`<schema-version>`) is ignored.

Also includes some preparation refactorings and some additional tests in separate commits. 

# Why this way

The refactorings are needed to avoid code duplication.

An alternative solution would be to allow querying the validation endpoint with no version (`/compatibility/subjects/<subject-key>`), and apply the fix only in that case.

### CODE REVIEW

File: website/source/quickstart.rst
Comment (davide-armand) [inline website/source/quickstart.rst:63]: Is there any other documentation that needs to be updated?

File: karapace/schema_registry.py
Comment (eliax1996) [inline karapace/schema_registry.py:355]: mmm isn't better to return all the errors? Maybe a string interpolation with all the errors its better than returning just the first

File: karapace/schema_registry.py
Comment (eliax1996) [inline karapace/schema_registry.py:482]: maybe:
```
            if is_incompatible(result):
                return result
```
more direct with the intent and less error prone in refactoring

File: karapace/schema_registry_apis.py
Comment (eliax1996) [inline karapace/schema_registry_apis.py:1322]: maybe private? It make sense to expose this outside of this file? I think its strictly bound to the webpage logic

File: tests/unit/test_schema_registry.py
Comment (eliax1996) [inline tests/unit/test_schema_registry.py:22]: Can you extract an interface and try to use DI, not blocking on this since its matter of taste, but I find hard to understand what this object its trying to do and what is the expected behaviour

File: tests/unit/test_schema_registry.py
Comment (eliax1996) [inline tests/unit/test_schema_registry.py:93]: got it, but why not creating a test where you know that the schemas are compatible 1 by 1 but not fully transitive? Isn't better to test the outcome instead of the actions? Again, matter of taste. Maybe better to ask a second opinion on this before doing work :D Maybe its me that isn't used to test like this and isn't anything bad

File: tests/unit/test_schema_registry.py
Comment (eliax1996) [inline tests/unit/test_schema_registry.py:141]: kinda the same comment, I would like to see two test. 1 where the schemas are compatible and the api reply green, one where the schemas are 2 by 2 compatible but not fully transitive compatible and the api reply red

**Review by eliax1996 — COMMENTED** [review]
I think the code its nice, the feature its well tested.
I'm still not fully convinced by the use of mocks, I think they can couple the way of a certain result its computed.
I would much more like a test where you give the input and the real implementation does the logic and you check the outcome, its a type of testing that makes the refactor much less painful since you don't need to change the assertions not being coupled by the names of the functions and the order in which they are called

File: karapace/schema_registry_apis.py
Comment (davide-armand) [reply karapace/schema_registry_apis.py]: Made it private, although it is used outside the class in a test though (`tests/unit/test_schema_registry_api.py`), does that count? :)
My Pyright complains about it, but maybe I should just remove that check in my setup

File: karapace/schema_registry.py
Comment (jjaakola-aiven) [reply karapace/schema_registry.py:355]: Yes, it would be better to hand out all errors instead of first.

File: karapace/schema_registry_apis.py
Comment (jjaakola-aiven) [inline karapace/schema_registry_apis.py:1343]:
```suggestion
    def _get_old_schema(self, subject: Subject, version: Version, content_type: str) -> ParsedTypedSchema:
```
The old schemas are not fully validated, there can be such old schemas in the Karapace database that cannot pass validation anymore with newer and current Avro library version.

File: tests/unit/test_schema_registry.py
Comment (jjaakola-aiven) [inline tests/unit/test_schema_registry.py:1]: Tests here could be integration tests and might be clearer there.

File: karapace/schema_registry.py
Comment (davide-armand) [reply karapace/schema_registry.py]: @jjaakola-aiven Good point, I have fixed it in a separate commit since not directly in the scope of this ticket.
I made a simple comma separated list, not sure if good enough.

I have also fixed it for the `/compatibility` endpoint here (it's json in this case):
https://github.com/Aiven-Open/karapace/pull/953/files#diff-a79ae23c164ad9d4d7ba51504e6d20f25dbbdc2b59f4d91bc2f105e3da9a6e63R410

File: tests/unit/test_schema_registry.py
Comment (davide-armand) [reply tests/unit/test_schema_registry.py:1]: @jjaakola-aiven I have move all test to `tests/integration/test_schema_compatibility.py`.
No mocks now and hopefully clearer than before :)

File: tests/unit/test_schema_registry.py
Comment (davide-armand) [reply tests/unit/test_schema_registry.py]: Moved to `tests/integration/test_schema_compatibility.py`

File: tests/unit/test_schema_registry.py
Comment (davide-armand) [reply tests/unit/test_schema_registry.py]: Moved to `tests/integration/test_schema_compatibility.py`

File: tests/unit/test_schema_registry.py
Comment (davide-armand) [reply tests/unit/test_schema_registry.py]: Moved to `tests/integration/test_schema_compatibility.py`

**Comment by davide-armand** [comment]
Rebased and fixed conflicts

File: website/source/quickstart.rst
Comment (eliax1996) [reply website/source/quickstart.rst:63]: I don't think so. The documentation should be revised completely IMO but for that PR I think its fine as it is. PS good addition

File: src/karapace/compatibility/schema_compatibility.py
Comment (eliax1996) [inline src/karapace/compatibility/schema_compatibility.py:55]: nit: why those empty lines between the `if` and `elif`?

File: src/karapace/schema_registry.py
Comment (eliax1996) [inline src/karapace/schema_registry.py:463]: nit: do we need the explicit type declaration here? 🤔
It should inferred by the return type of the function

File: src/karapace/schema_registry_apis.py
Comment (eliax1996) [inline src/karapace/schema_registry_apis.py:410]: again super nit, I would declare the error message as a separate variable like:
```python3
maybe_truncated_error = ', '.join(result.messages)[:300]
self.r({"is_compatible": False, "incompatibilities": maybe_truncated_error}, content_type)
```

File: tests/integration/test_dependencies_compatibility_protobuf.py
Comment (eliax1996) [inline tests/integration/test_dependencies_compatibility_protobuf.py:186]: issue: not its true also for `None` that its the default value of the `.get` when its empty

File: tests/integration/test_dependencies_compatibility_protobuf.py
Comment (eliax1996) [inline tests/integration/test_dependencies_compatibility_protobuf.py:274]: same, I would change it tto: `assert res.json().get("is_compatible") is False`

File: tests/integration/test_dependencies_compatibility_protobuf.py
Comment (eliax1996) [inline tests/integration/test_dependencies_compatibility_protobuf.py:342]: same

File: tests/integration/test_dependencies_compatibility_protobuf.py
Comment (eliax1996) [inline tests/integration/test_dependencies_compatibility_protobuf.py:410]: and here

File: tests/integration/test_dependencies_compatibility_protobuf.py
Comment (eliax1996) [inline tests/integration/test_dependencies_compatibility_protobuf.py:494]: here

File: tests/integration/test_schema_compatibility.py
Comment (eliax1996) [inline tests/integration/test_schema_compatibility.py:21]: nit: add `Final` as type

File: tests/integration/test_schema_compatibility.py
Comment (eliax1996) [inline tests/integration/test_schema_compatibility.py:28]: optional: wdyt of creating a type alias for this? Not required, just asking if make sense in your opinion

File: tests/integration/test_schema_compatibility.py
Comment (eliax1996) [inline tests/integration/test_schema_compatibility.py:29]: maybe `bool | None`?
If you add at the beginning: `from __future__ import annotations` so you can use also the type annotations of the most recent python in the old versions

File: tests/integration/test_schema_compatibility.py
Comment (eliax1996) [inline tests/integration/test_schema_compatibility.py:31]: same here, but should be fixed by the `pre-commit` as soon as you import the `annotations`

File: tests/integration/test_schema_compatibility.py
Comment (eliax1996) [inline tests/integration/test_schema_compatibility.py:194]: awesome! Thanks, love this style 🥳 🚀

**Review by eliax1996 — COMMENTED** [review]
Almost there, except for a couple of changes great great job!
I like much more those tests, I can read them and understand them really well, thanks for the hard work 🚀

File: src/karapace/compatibility/schema_compatibility.py
Comment (davide-armand) [reply src/karapace/compatibility/schema_compatibility.py]: Not sure, I think it was pre-existing code that I moved around.
I have now removed the empty lines so it's more compact :)

File: src/karapace/schema_registry_apis.py
Comment (davide-armand) [reply src/karapace/schema_registry_apis.py]: Agree, looks more readable now!

File: tests/integration/test_dependencies_compatibility_protobuf.py
Comment (davide-armand) [reply tests/integration/test_dependencies_compatibility_protobuf.py]: Good catch, fixed

File: tests/integration/test_schema_compatibility.py
Comment (davide-armand) [reply tests/integration/test_schema_compatibility.py]: Agree, more readable

File: tests/integration/test_schema_compatibility.py
Comment (davide-armand) [reply tests/integration/test_schema_compatibility.py]: Didn't know that :D
Fixed

File: karapace/schema_registry_apis.py
Comment (davide-armand) [reply karapace/schema_registry_apis.py]: Good catch, fixed

**Comment by davide-armand** [comment]
@eliax1996 I should I have address all comments, thanks for your detailed review!

File: tests/integration/test_schema.py
Comment (eliax1996) [reply tests/integration/test_schema.py:335]: mmm maybe you forgot this

**Review by eliax1996 — APPROVED** [review]

**Pull request merged**
